### DISCUSSION TITLE
iLogtail observability enhancement

### DISCUSSION THREAD

**Discussion by linrunqi08** [alibaba/ilogtail#716] [Ideas]
Current Status：https://ilogtail.gitbook.io/ilogtail-docs/configuration/logging

1. Support Exporting ilogtail_profile and ilogtail_status by HTTP（promethus）
2. Improve C++partial monitoring data
3. Unified C++and Golang monitoring data

**Comment by linrunqi08** [comment]
<div class="lake-content" typography="classic"><h2 id="ymRYm" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">iLogtail自身可观测性现状</span></h2><p id="ua0947c2c" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">iLogtail当前对外可见可观测指标通过snapshot目录下的ilogtail_status.LOG文件和ilogtail_profile.LOG文件对外暴露：</span></p><p id="ua8751b07" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">ilogtail_status.LOG：全局统计指标</span></p>

资源消耗 | CPU、MEM
-- | --
配置 | 配置数量、更新时间等
文件读取 | 打开文件数、监控目录数、读取次数等
处理 | 处理队列大小、处理行数、字节数等
发送 | 发送次数、字节数、异常发送队列数等

<p id="u4da868c4" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">当前自身可观测指标有这样几个问题：</span></p><ol class="ne-ol" style="margin: 0; padding-left: 23px"><li id="u5ac0b25d"><span class="ne-text">指标缺少明确定义，不知道count是次数还是条数，不知道统计周期等，指标很多但看不懂</span></li><li id="u82eaf358"><span class="ne-text">细粒度指标文件类和其他类型差别太大，不仅其他类型指标远少于文件采集，且即使相同指标口径也未必一致，不利于问题排查</span></li><li id="u1438f59b"><span class="ne-text">细粒度指标时间粒度不可调整</span></li><li id="uf638d350"><span class="ne-text">缺少对数据流处理的洞察指标，例如每个配置，各个处理插件的统计指标，处理流程不透明</span></li><li id="u62555f41"><span class="ne-text">指标缺少对外暴露能力，如通过Prometheus Format被第三方监控抓取或原生作为一种指标Input被采集处理发送到第三方存储</span></li><li id="ub1bfe577"><span class="ne-text">实现上不同组件自由实现，相关代码散乱在各处，新增模块添加指标无从下手</span></li></ol><h2 id="i84HX" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">期望的可观测性</span></h2><ul class="ne-ul" style="margin: 0; padding-left: 23px"><li id="uf35c39e4"><span class="ne-text">统一：指标具有明确定义，且形成文档和命名规范便于理解。文件类（C++）和其他类型（Go）指标统计口径统一。</span></li><li id="udc657c7d"><span class="ne-text">丰富：细粒度的每个插件实例的指标清晰展示，分为公共部分和特有部分，具备足够信息可视化还原处理流水线状态。</span></li><li id="u7ceea35b"><span class="ne-text">可导出：可以被Prometheus被抓取或原生作为一种指标Input被采集处理发送到第三方存储</span></li><li id="uf86e80bc"><span class="ne-text">可扩展：实现上具备良好的指标采集框架，新增插件可方便接入</span></li></ul><h2 id="p9DGz" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">长期目标</span></h2><h3 id="Cv42g" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">定义公共的指标和命名规范</span></h3><p id="u7e1e8435" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">采集器资源状态</span></p><pre data-language="plain" id="fopxf" class="ne-codeblock language-plain" style="border: 1px solid #e8e8e8; border-radius: 2px; background: #f9f9f9; padding: 16px; font-size: 13px; color: #595959">cpu
mem
fd_open</pre><p id="u6a1ce206" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">Input</span></p><pre data-language="plain" id="D3Lng" class="ne-codeblock language-plain" style="border: 1px solid #e8e8e8; border-radius: 2px; background: #f9f9f9; padding: 16px; font-size: 13px; color: #595959">input_records_out_total
input_bytes_out_total</pre><p id="uf93bac9e" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">Processor</span></p><pre data-language="plain" id="Wx0Ay" class="ne-codeblock language-plain" style="border: 1px solid #e8e8e8; border-radius: 2px; background: #f9f9f9; padding: 16px; font-size: 13px; color: #595959">process_record_in_total
process_bytes_in_total
process_record_out_total
process_bytes_out_total</pre><p id="u8aa69c82" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">Flusher</span></p><pre data-language="plain" id="w4XFj" class="ne-codeblock language-plain" style="border: 1px solid #e8e8e8; border-radius: 2px; background: #f9f9f9; padding: 16px; font-size: 13px; color: #595959">flush_record_in_total
flush_bytes_in_total
flush_record_out_total
flush_bytes_out_total
flush_count_net_out_total
flush_bytes_net_out_total
flush_record_success_total
flush_bytes_success_total
flush_record_retried_total
flush_bytes_retried_total
flush_record_dropped_total
flush_bytes_dropped_total</pre><p id="ua8668b41" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"></p><h3 id="nNtlA" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">层次（label）</span></h3><p id="u198b1112" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">1 全局：hostname=&quot;xx&quot;、version=&quot;xx&quot;、os=&quot;xx&quot;</span></p><p id="u6d904917" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">2 插件实例：name=&quot;plugin_type/plugin_id&quot;</span></p><p id="u38fbcf87" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">3 配置：config=&quot;xx&quot;</span></p><p id="u08e36ee8" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">4 采集对象实例：filename=&quot;xxx&quot;、upstream=&quot;xxx&quot;</span></p><p id="uc973c78d" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text"></span></p><h3 id="hTmxG" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">指标类型</span></h3><ul class="ne-ul" style="margin: 0; padding-left: 23px"><li id="ua5197593"><span class="ne-text" style="color: rgb(51, 51, 51); font-size: 16px">Counter</span></li><li id="ud79108aa"><span class="ne-text" style="color: rgb(51, 51, 51); font-size: 16px">Gauge</span></li><li id="u1c2bcbf7"><span class="ne-text" style="color: rgb(51, 51, 51); font-size: 16px">Summary</span></li></ul><h3 id="qkbvA" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">Exporter和导出插件</span></h3><p id="u1209142a" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><img src="https://intranetproxy.alipay.com/skylark/lark/0/2023/png/31056853/1677040258920-fec3ac74-8149-4ea7-99e0-c0b56167838f.png" width="774.5454377576343" id="u5574e148" class="ne-image"></p><h3 id="zSmES" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">Metric相关接口</span></h3><h4 id="p36ne" style="font-size: 16px; line-height: 24px; margin: 10px 0 5px 0"><span class="ne-text">Plugin接口</span></h4><p id="uafec18c8" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">GetMetrics</span></p><h4 id="dYnId" style="font-size: 16px; line-height: 24px; margin: 10px 0 5px 0"><span class="ne-text">Metrics接口</span></h4><p id="u11a37500" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text" style="font-size: 16px">CreateCounter</span></p><p id="ue7e4f579" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text" style="font-size: 16px">CreateGauge</span></p><p id="ufb8ad701" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text" style="font-size: 16px">用此类接口创建Metric，成员变量可以直接访问具体Metric</span></p><p id="uabcf62b4" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">ForEachCounter</span></p><p id="u129ee34c" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">ForEachGauge</span></p><p id="u6ac9bbea" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">用此类接口遍历Metric，无论自定义多少Metric都可以全部遍历</span></p><p id="u95f33228" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><img src="https://intranetproxy.alipay.com/skylark/lark/0/2023/png/31056853/1677043949617-0583de3b-c60f-46db-aa49-9ef8c99a6c48.png" width="289.9999937144194" id="u20b9ca30" class="ne-image"></p><h4 id="wxLAT" style="font-size: 16px; line-height: 24px; margin: 10px 0 5px 0"><span class="ne-text">Metric接口</span></h4><p id="ubb931478" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text" style="font-size: 16px">Gauge</span></p><p id="u57161d6a" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Set</span></p><p id="u417e8ca6" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Add</span></p><p id="u820c8288" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Inc</span></p><p id="u5a8417a6" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Sub</span></p><p id="ucf068b64" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Dec</span></p><p id="ucaa148df" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Get</span></p><p id="u10ce2b6d" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text" style="font-size: 16px">Counter</span></p><p id="u6b21f74b" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Set</span></p><p id="u0b3bc778" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Add</span></p><p id="u6c033657" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Inc</span></p><p id="u690d9998" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; margin-left: 2em"><span class="ne-text" style="font-size: 16px">Get</span></p><p id="u6985e6b0" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text" style="font-size: 16px">Summary</span></p><p id="u551cca1a" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text" style="font-size: 16px"></span></p><h2 id="X8bT1" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">短期怎么做</span></h2><p id="u2a8cd231" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text" style="font-size: 16px">C++当前实现</span></p><p id="ubbae5256" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><img src="https://intranetproxy.alipay.com/skylark/lark/0/2023/png/31056853/1677046194531-9c7cbf92-d0f2-4092-bf2d-da0185eedfce.png" width="804.5454371074018" id="ua55dda25" class="ne-image"></p><p id="uf3f4e09b" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">Go当前实现</span></p><p id="u953ff42c" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><img src="https://intranetproxy.alipay.com/skylark/lark/0/2023/png/31056853/1677064060361-099a023c-d053-4071-9d34-d0eabb6db6c6.png" width="448.1818084677391" id="u77443863" class="ne-image"></p><p id="ub5bb254a" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">使用统一数据结构改造，无锁化。Go复用v2的metric模型。</span></li><li id="ucd53f6b1"><span class="ne-text">C++保留一个集中的InternalMetrics收集类，使用插件注册的形式告知其要收集数据的模块实例</span></li><li id="ue043b9e5"><span class="ne-text">文件dump功能作为消费模块独立</span></li><li id="ua7645e65"><span class="ne-text">MetricExporter和MetricInput复用Go现有框架，暂时通过插件暴露指标。</span></li><li id="u2b604197"><span class="ne-text">C++侧指标通过cgo插件允许Go访问。</span></li></ul><p id="u23e5a9ab" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><img src="https://intranetproxy.alipay.com/skylark/lark/0/2023/png/31056853/1677060001958-f14d9f75-1768-4c65-8b60-ab86d107929b.png" width="770.9090742000867" id="u7d9de75c" class="ne-image"></p><p id="u6edd1bd9" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"></p></div>

**Reply by wangpeix** [reply]
采集延迟指标也是需要的，用来衡量采集延迟多久。如采集延迟时间...

**Reply by messixukejia** [reply]
收到，这个在考虑了。

**Comment by linrunqi08** [comment]
Golang
![image](https://github.com/alibaba/ilogtail/assets/90741255/1ac423f1-66ed-4673-a1ed-baef51679b1e)

每个plugin会hold一个metric对象，LogstoreConfig会hold一个metric对象的列表，在加载每个plugin的时候，将Metric对象放到列表中，这样每个插件只对自己的metric对象进行操作，不需要锁

C++
![image](https://github.com/alibaba/ilogtail/assets/90741255/237d5d16-f43a-4bf5-9e06-7b3abe24cf34)
支持Push or Pull
● [Pull or Push？监控系统如何选型](https://developer.aliyun.com/article/788460?utm_content=m_1000292537)）
● Pull模式在开源场景下基本上可以说是行业标准

**Reply by wangpeix** [reply]
指标支持主动推送到下游的kafka吗。
有几万台机器， 每台机器上一个logtail，指标支持推送到下游kafka，方便下游统一处理

**Reply by messixukejia** [reply]
后面是可以考虑下作为一种数据采集到任意下游 @linrunqi08

**Comment by cyenxchen** [comment]
@linrunqi08 老哥有些图挂了，方便补一下吗？

**Comment by linrunqi08** [comment]
```
// 基础指标类
class BaseMetric{
    struct MetricObj {
        /* counters and gauges */
        std::atomic_long val;

        std::atomic_long timestamp;
    };
    public:
        BaseMetric();
        BaseMetric(MetricObj* metricObj);
        ~BaseMetric();
        MetricObj* mMetricObj;
        
        // 指标操作 
        void baseMetricAdd(uint64_t val);
        MetricObj* getMetricObj();
        MetricObj* snapShotMetricObj();
};

// Pipeline指标，对基础指标进行封装
class PipelineMetric {
    public:
        // 指标名->指标的Map
        std::unordered_map<std::string, BaseMetric*> mBaseMetrics;
        // 标签，层级使用标签来区分
        std::unordered_map<std::string, std::string> mLabels;
};
```

**Reply by linrunqi08** [reply]
https://github.com/alibaba/ilogtail/pull/914

**Comment by linrunqi08** [comment]
需求
允许增加、删除指标项（相同Label组合认为一项），但不是频繁操作，一般切文件才会增删
修改指标值，频繁操作，不能被block
读取指标，不是频繁操作，一般几秒一次
读写分离设计
gWriteMetrics用于被写
gReadMetrics用于被读
定期将gWriteMetrics快照到gReadMetrics中，同时重置gWriteMetrics中指标，清理gWriteMetrics中无效指标项
写Metrics设计
gWriteMetrics全局唯一
使用方使用CreateMetrics接口创建一个Metrics，接口形式为Metrics* CreateMetrics(Labels)
Metrics的Label不支持修改
Metrics可以增加具体指标值，通过XXX* Metrics::CreateXXX(Name)，XXX可以为Counter、Gauge等。
使用方通过在自己类内记录下这些WeatPtr直接修改指标值，无查找指标需求。
例如：
```
class Gauge {
    atomic_long mTimestamp;
    atomic_long mValue;
    void Set(long val) {
        mTimestamp = time();
        mValue = val;
    }
};
class Metrics {
    vector<pair<string>> mLabels;
    List mValues;
    Gauge* CreateGauge(string Name);
    Metrics* prev, next;
};

class WriteMetrics {
    Metrics* CreateMetrics(vector<string> Labels);
};

WriteMetrics* gWriteMetrics;

class Reader {
    bool Init() {
        vector<string> labels = {
            {"Config","config1"},            
            {"Path","path1"}, 
        };
        mMetrics = gWriteMetrics->CreateMetrics(labels);
        mOffsetGauge = mMetrics->CreateGauge("Offset");
    }
    ~Reader() {
        gWriteMetrics.Destroy(mMetics);
    }
    bool ReadLines() {
        mOffsetGause->Set(offset);
    }
    WeakPtr<Metrics> mMetrics;
    WeakPtr<Gauge> mOffsetGauge;
}

```
读Metics设计
gReadMetrics是全局唯一，可以具备自己的数据结构和方法，为读取优化，这里重点考虑数据怎么来。
gReadMetrics中的数据从gWriteMetrics中通过快照得到，因此需要能遍历gWriteMetrics中的数据。遍历的时候希望不要影响Metrics的写入，同时不要因指标项删除发生异常。因此考虑使用DoubleLinkedList数据结构存储Metrics，一写多读安全，同时使用延时删除技术避免指标项删除对读取到影响。
```
class WriteMetrics {
    Metrics* CreateMetrics(vector<string> Labels) {
      lock_guard(mMutex);
      // link new Metrics to mHead and return;
    }
    void DestroyMetrics(Metrics* metrics) {
      lock_guard(mMutex);
      // unlink metrics from mHead and add to mDeletedMetrics;
      // do not modify metrics's next pointer to let who is reading this Metrics can read next Metrics normally
    }
    void clearDeleted() {
      {
        lock_guard(mMutex);
        vector<Metrics*> toDelete;
        toDelete.swap(mDeletedMetrics);
      }
      // real delete pointers in toDelete
    }
    void DoSnapshot(Metrics*& snapshot) {
      for (Metrics* metrics = mHead; metrics != mTail; metrics=metrics->next) {
        // copy/swap metrics to snapshot
      }
      clearDeleted();
    }
    Metrics* mHead;
    Metrics* mTail;
    vector<Metrics*> mDeletedMetrics;
    mutex mMutex;
};
```
gReadMetrics没有实时写入的需求，但在做snapshot更新后，还是需要保证当前正在读的人正常因此可以简单加一下锁。
```
class ReadMetrics {
    void ReadMetrics(……) {
      rlock_guard(mMutex);
      // Do some read
    }

    void UpdateMetrics() {
      Metrics* snapshot;
      gWriteMetrics.DoShapshot(snapshot);
      Metrics* toDelete = mHead;
      {
        wlock_guard(mMutex);
        mHead = snapshot;
      }
      // do deletion
    }
    Metrics* mHead;
    shared_mutex mMutex;
};

class ReadMetricsIterater {
  ReadMetricsIterater() {
    rlock(gReadMetrics->mMutex);
  }
  ~ReadMetricsIterater() {
    unrlock(gReadMetrics->mMutex);
  }
}
```

**Comment by linrunqi08** [comment]
![image](https://github.com/alibaba/ilogtail/assets/90741255/30b86118-c7ac-4168-8f74-8f189e669b67)

**Reply by linrunqi08** [reply]
mDeletedMetrics 去掉，采用标记删除的方式。在do snapshot的时候，边复制，边删除

**Comment by cyenxchen** [comment]
@linrunqi08  有个问题，现在的设计是c++部分和go插件部分产生的指标是分开来送出去吗？还是c++产生的指标先送到go插件系统里然后在统一送出去？

**Reply by messixukejia** [reply]
会汇集的，框架代码最近会入库，可以关注下。

如果有兴趣，可以一起来开发下发送到第三方系统的部分

**Reply by cyenxchen** [reply]
可以的

**Comment by liuhaoyang** [comment]
目前这个feature是什么进展了

**Reply by linrunqi08** [reply]
代码已合入，可以看下这几个pr：
https://github.com/alibaba/ilogtail/pull/1015
https://github.com/alibaba/ilogtail/pull/1106
https://github.com/alibaba/ilogtail/pull/1167

**Comment by linrunqi08** [comment]
最新进展，可以看下pr: https://github.com/alibaba/ilogtail/pull/1290   目前支持golang 部分 v1 pipeline完整指标统一通过Cpp exporter输出。

**Reply by linrunqi08** [reply]
目前Cpp Exporter需要实现Promethus API从而将指标暴露出来。希望有人认领一下改功能，比较简单

**Comment by linrunqi08** [comment]
当前状态:
整体：
![image](https://github.com/alibaba/ilogtail/assets/90741255/6e29e4c7-66e7-4c53-aa9a-3421038895ae)

C++
![image](https://github.com/alibaba/ilogtail/assets/90741255/59c1a494-8207-4337-a541-2eb45f96c57c)


如何为插件添加自定指标？
https://github.com/alibaba/ilogtail/pull/1379/files#diff-7e45e30fb6876b0f28b367e761d31d38e7adadc8dda84ced6211d1987708ec13

**Comment by linrunqi08** [comment]
开发者招募：

在C++中实现HttpServer，以Prometheus 格式的API方式对外暴露指标数据

部分改动参考接口定义：
https://github.com/alibaba/ilogtail/tree/feature/metric_exporter

**Comment by linrunqi08** [comment]
关键PR
1. Golang插件指标重构代码： https://github.com/alibaba/ilogtail/pull/1290/files
2. C++插件指标重构代码： https://github.com/alibaba/ilogtail/pull/1015/files
3. C++插件添加指标demo：https://github.com/alibaba/ilogtail/pull/1106/files
